### DISCUSSION TITLE
Updating external models in Langium correctly

### DISCUSSION THREAD

**Discussion by bjthehun** [eclipse-langium/langium#1141] [Q&A] [answered]
I have integrated an external model into Langium as described [here](https://github.com/langium/langium/discussions/984), and so far I've succeeded in extracting all references I need for the user-facing language. 
However, the user may update the external model, so I need to update the references in Langium automatically as well. This doesn't work properly for me right now.

Whenever my language server receives an update in the external model, I reconstruct the references like this:
```ts
let ws = services.sharedServices.workspace;
const referenceAST = extractReferencesToAST(externalModel)

let referencesDocument = 
  ws.LangiumDocumentFactory.fromModel(referenceAST, pathToRefs);
await ws.DocumentBuilder.build(
  [referencesDocument], 
  {validationChecks: 'all'}
);

if (ws.LangiumDocuments.hasDocument(pathToRefs)) {
  ws.LangiumDocuments.deleteDocument(pathToRefs);
}    
ws.LangiumDocuments.addDocument(referencesDocument);
```
However, with this code, I miss the last update, as the newly calculated references don't show up until I receive the next update. So what am I doing wrong?

**Comment by msujew** [comment]
Hey @bjthehun,

I assume the way you update the document is incorrect. The document builder probably takes the old document instead of the updated one, which would explain why it's always lagging one update cycle behind. The following should work better:

```ts
let ws = services.sharedServices.workspace;
const referenceAST = extractReferencesToAST(externalModel)

let referencesDocument = 
  ws.LangiumDocumentFactory.fromModel(referenceAST, pathToRefs);
if (ws.LangiumDocuments.hasDocument(pathToRefs)) {
  ws.LangiumDocuments.deleteDocument(pathToRefs);
}    
ws.LangiumDocuments.addDocument(referencesDocument);
await ws.DocumentBuilder.update(
  [referencesDocument.uri], 
  []
);
```

**Reply by bjthehun** [reply]
Unfortunately, I instead get this error:
```
Error: Method not implemented.
      at EmptyFileSystemProvider.readFileSync (node_modules/langium/src/workspace/file-system-provider.ts:45:15)
      at DefaultLangiumDocumentFactory.getContentFromFileSystem (node_modules/langium/src/workspace/documents.ts:246:40)
      at DefaultLangiumDocumentFactory.update (node_modules/langium/src/workspace/documents.ts:222:67)
      at ~/project/node_modules/langium/src/workspace/document-builder.ts:137:41
      at DefaultDocumentBuilder.runCancelable (node_modules/langium/src/workspace/document-builder.ts:166:19)
      at async DefaultDocumentBuilder.buildDocuments (node_modules/langium/src/workspace/document-builder.ts:136:9)
      at async DefaultDocumentBuilder.update (node_modules/langium/src/workspace/document-builder.ts:102:9)
...
```

What I forgot to mention is I hold the external model in memory (`pathToRefs` starts with `memory://...`), and therefore, I'm using `EmptyFileSystem` as `FileSystemProvider`.

**Reply by bjthehun** [reply]
From what I can see, Langium tries to parse my model references again, even though I create an AST without parsing, following [this suggestion](https://github.com/langium/langium/discussions/984#discussioncomment-5416442).

But even when I want to parse the external model as text document, `update` in `DefaultLangiumDocumentFactory` ignores the `textDocument` property of the `document` it gets as input. Is that intended?

**Reply by msujew** [reply]
Right, I forgot that `update` will actually discard the AST and reattempt parsing the model. Note that it doesn't ignore the `textDocument`, but will attempt to get the latest text from the `TextDocuments` service which only works in case the document is actually opened in an editor. You can try continue using `DocumentBuilder.build` though. That should likely work.

**Reply by bjthehun** [reply, marked as answer]
I've figured out that `build` and `update` work in combination for me:

```ts
let ws = services.sharedServices.workspace;
const referenceAST = extractReferencesToAST(externalModel)

let referencesDocument = 
    ws.LangiumDocumentFactory.fromModel(referenceAST, pathToRefs);

if (ws.LangiumDocuments.hasDocument(pathToRefs)) {
    ws.DocumentBuilder.update([], [pathToRefs]);
} 
await ws.DocumentBuilder.build(
    [referencesDocument], 
    {validationChecks: 'all'}
);
ws.LangiumDocuments.addDocument(referencesDocument);
```

So once I tell `DocumentBuilder` i want to remove the existing external model, the "update" succeeds. 

Thanks for the tip with `build`.
